### PR TITLE
Add support for config through js

### LINKED ISSUE #183: Configuration through JS
Here is a proposal to allow configuration of zuul through JS.

Let there be `zuul.config.js` which takes precedence over `zuul.yaml` (we should warn if both exist though). The file is a CommonJS module which exports zuul config as an object:
```
module.exports = {
  ui: 'mocha',
  browsers: [
    {name: 'chrome',
     version: 'latest'},
    {name: 'safari',
     version: 'latest'}
  ],
  ...
}
```

Pros:
* Configuration based on environment variables
* Configuration hooks for builders (a callback which configures browserify, so we don't have to track its API closely)
* Using JS for functions/modules (for example if we need to maintain multiple test configurations)
* JS is more familiar than YAML

Cons:
* It's executable code, there can be additional security risks... (not sure how much we concerned with that though given that a lot of tools do the same)
* Not as much declarative as yaml (what does it even mean?)

### PR DESCRIPTION
**Pull request by andreypopp**
We add support for local configuration through JS by looking for a CommonJS
module named `zuul.config.js`.

We exit if we encounter both `.zuul.yml` and `zuul.config.js` as porting from
one to another is straightforward and we don't want to introduce possible
confusion by merging them.

Closes #183.

### CODE REVIEW

**Comment by andreypopp** [comment]
Sorry, messed up with branches and removed the one which belongs to PR accidentally. Please consider this PR.

**Comment by andreypopp** [comment]
Any thoughts on this?

**Comment by defunctzombie** [comment]
I'll review this weekend.

**Comment by vvo** [comment]
I like it, this looks good to me

**Comment by vvo** [comment]
@andreypopp If this get merged you will also need to update the wiki if that's feasible

File: bin/zuul [hunk 1]
```diff
@@ -333,4 +312,42 @@ scout_browser(function(err, all_browsers) {
     });
 });
 
+function readLocalConfig(config) {
+    var yaml = path.join(process.cwd(), '.zuul.yml')
+    var js = path.join(process.cwd(), 'zuul.config.js')
+    var yamlExists = fs.existsSync(yaml)
+    var jsExists = fs.existsSync(js)
+    if (yamlExists && jsExists) {
+        console.error('both .zuul.yaml and zuul.config.js are found in the project directory')
```
Comment (vvo) [inline bin/zuul]: "Please choose one" => helps the user solving the issue

File: bin/zuul [hunk 2]
```diff
@@ -333,4 +312,42 @@ scout_browser(function(err, all_browsers) {
...
+        return mergeConfig(config, readYAMLConfig(yaml))
+    } else if (jsExists) {
+        return mergeConfig(config, require(js))
+    } else {
+        return config
+    }
+}
+
+function readGlobalConfig(config) {
+    var filename = find_nearest_file('.zuulrc') || path.join(osenv.home(), '.zuulrc')
```
Comment (vvo) [inline bin/zuul]: So it means you cannot have a global config written in JavaScript? An so we need to document both YAML global and local JS. Could we have global JS too, would it make sense?

Comment (andreypopp) [inline bin/zuul]: Good point... what do you think should be the name of file within the home directory? The same `.zuulrc`? We could try to parse it with YAML and then fallback to JS.

**Comment by andreypopp** [comment]
I've addressed your comments, @vvo

**Comment by andreypopp** [comment]
rebased on current master

**Comment by andreypopp** [comment]
Would love it to be merged soon as we already use such functionality — https://github.com/prometheusresearch/zuul-builder-webpack

**Comment by vvo** [comment]
Using this PR on my project, I got:

```sh
/home/vvo/.zuulrc:2
sauce_key: xxxx
           ^^^^
```

Like an error showing up while parsing my zuulrc. Used:

```sh
zuul --phantom -- test/run-browser.js
```

**Comment by defunctzombie** [comment]
Is this PR good to go or somehow broken still @vvo ?

File: bin/zuul [hunk 3]
```diff
@@ -333,4 +312,48 @@ scout_browser(function(err, all_browsers) {
...
+    if (fs.existsSync(filename)) {
+        var globalConfig;
+        try {
+            globalConfig = require(filename);
+        } catch(_err) {
```
Comment (defunctzombie) [inline bin/zuul]: why the underscore?

File: bin/zuul [hunk 4]
```diff
@@ -333,4 +312,48 @@ scout_browser(function(err, all_browsers) {
...
+            globalConfig = readYAMLConfig(filename);
+        }
+        return mergeConfig(config, globalConfig);
+    } else {
```
Comment (defunctzombie) [inline bin/zuul]: don't need the else here because of the return in the `if`

File: bin/zuul [hunk 5]
```diff
@@ -333,4 +312,48 @@ scout_browser(function(err, all_browsers) {
...
+    var yamlExists = fs.existsSync(yaml)
+    var jsExists = fs.existsSync(js)
+    if (yamlExists && jsExists) {
+        console.error('both .zuul.yaml and zuul.config.js are found in the project directory, please choose one')
+        process.exit(1)
+    } else if (yamlExists) {
+        return mergeConfig(config, readYAMLConfig(yaml))
+    } else if (jsExists) {
+        return mergeConfig(config, require(js))
+    } else {
```
Comment (defunctzombie) [inline bin/zuul]: don't need the else here because of the return in the `if`

File: bin/zuul [hunk 6]
```diff
@@ -333,4 +312,48 @@ scout_browser(function(err, all_browsers) {
...
 
+function readLocalConfig(config) {
+    var yaml = path.join(process.cwd(), '.zuul.yml')
+    var js = path.join(process.cwd(), 'zuul.config.js')
+    var yamlExists = fs.existsSync(yaml)
+    var jsExists = fs.existsSync(js)
+    if (yamlExists && jsExists) {
+        console.error('both .zuul.yaml and zuul.config.js are found in the project directory, please choose one')
+        process.exit(1)
+    } else if (yamlExists) {
```
Comment (defunctzombie) [inline bin/zuul]: style is:

```js
if () {
}
else {
}
```

**Comment by defunctzombie** [comment]
Just some minor cosmetic things, otherwise seems good to me.

File: bin/zuul (hunk 3 above)
Comment (vvo) [inline bin/zuul]: I guess to annotate that's its not used, like I saw in golang code (and perhaps more)

**Comment by vvo** [comment]
> Is this PR good to go or somehow broken still @vvo ?

Not broken but there's something strange with the console message I get, any hint @andreypopp

**Comment by vvo** [comment]
@andreypopp did you see my comment about the faiure I got using your branch? let me know

**Comment by andreypopp** [comment]
Saw it, thanks, will investigate as I'm back to computer

File: bin/zuul (hunk 6 above)
Comment (andreypopp) [inline bin/zuul]: Hm... I actually see it's not enforced even within this file. Will fix my code but won't do it for other places not to pollut PR with unrelated changes.

File: bin/zuul (hunk 3 above)
Comment (andreypopp) [inline bin/zuul]: Yes, `catch()` is invalid syntax. Do you want it to be fixed?

File: bin/zuul (hunk 5 above)
Comment (andreypopp) [inline bin/zuul]: Prefer else for symmetry...

**Comment by andreypopp** [comment]
@vvo hm... actually I can't reproduce console message you reported. You have `.zuulrc` written in YAML? Can you try removing `node_modules/` and re-installing deps?

**Comment by defunctzombie** [comment]
This PR looks good to me. @andreypopp anything else you wanted to do here?

**Comment by andreypopp** [comment]
I think I'm done

**Comment by vvo** [comment]
> You have .zuulrc written in YAML? Can you try removing node_modules/ and re-installing deps?

Yes, my zuulrc is in YAML and when requiring this file in `globalConfig = require(filename);` it throws a global console.log

**Comment by vvo** [comment]
@andreypopp the require global console log only happens in node 0.10 when this:

```js
try {
  require('/fsa/fsa/f/saf/as/fsa');
} catch(e) {

}
```

So no big deal.
